fix some clippy suggestions

hellow554 opened the pull request:
This contains mostly using of [`cast`](https://doc.rust-lang.org/std/primitive.pointer.html#method.cast) instead of `as *const _` and the second thing is using the [`addr_of!` macro](https://doc.rust-lang.org/stable/std/ptr/macro.addr_of.html) for easier reading and avoiding possible UB (which isn't the case in any of the places I replaced the code).

hellow554 committed:
fix some clippy suggestions

fpagliughi commented:
Hey, thanks so much for the PR!

Unfortunately, I don't think I can merge it as-is right now. I wasn't clear about the Minimum Supported Rust Version on the README, but I personally still need to support rustc v1.46 for a production system. (Sorry, that's my mistake). And I wouldn't want to push the required version forward just for readability issues.

It looks like `addr_of_mut!()` requires v1.51. Most of the other stuff looks good, and useful, so maybe we can split the difference and get most of these changes in.

BTW: What clippy lint did you use? My defaults didn't catch most of these things.

hellow554 committed:
add clippy toml with msrv of 1.46.0

hellow554 commented:
I typically use `cargo clippy -- -W clippy::pedantic -W clippy::nursery`.
It gives a lot of warnings, especially docs, `#[must_use]` and const related stuff. But you can "disable" those with `-A clippy::must_use_candidate` for example. Give it a try ;)

I added a clippy toml and added the msrv of 1.46.0 as well as removed the `addr_of!` macro. Hope this is better :)

fpagliughi commented:
Yeah. That sounds great. I will have a look over the weekend and get it merged. Thanks again!
